### DISCUSSION TITLE
Can someone clarify the interaction between autocommit, autobegin, and actual transaction behaviors in SQLA?

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7126] [Usage Questions]
I find the SQLA documentation about autocommit and legacy vs new behaviors related to it, and the concept of "autobegin" very confusing. It is not at all clear what I should expect SQLA to do in various scenarios. Particularly for me, I'm using psycopg2 ...

First, my understanding is that with "autocommit=True" we expect that every statement issued to the DBAPI will be implicitly committed and from the SQLA documentation at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit it seems that with this mode enabled, the DBAPI then does not begin transactions under any circumstance -- presumably because every statement is an implicit commit anyway.

In my sessions I have autocommit=False and I have a  class that has relationships with lazy='joined'

As I understand the documentation at the above URL, under SQLA 1.4 we end up with transactions that are begun by SQLA wherever it deems appropriate, but what is not clear is whether those sessions are ever committed by SQLA itself???

Particularly, what commit behavior should I expect following an access to a relationship attribute, especially when that access emits SQL? Should I be expecting that any object graph traversals may have opened a transaction in the session to which the traversed objects belong?! In other words, is it idiomatic to do something like:

```
  with session.begin():
    object.other_object
```

That is, if I access some_object.other_object  where other_object = relationship(...,lazy='joined') then SQLA will emit the SQL to fetch the related other_object and I should expect that the session is then left in an open transaction? 

This questions is related to solving issues I'm asking about also here: https://github.com/sqlalchemy/sqlalchemy/discussions/7109

**Comment by zzzeek** [comment]
I think it's important here, first if you haven't already, to familiarize with the general Session usage pattern using the ORM tutorials. We have a new one for 1.4 and forward which shows the typical Session conversation at https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html .  The tutorial for the 1.x versions is also just as fine here, as nothing has really changed at all in how Session is used, so the sequence at https://docs.sqlalchemy.org/en/14/orm/tutorial.html#adding-and-updating-objects also shows the general idea.

the next thing is, when trying to understand and learn the Session, is to definitely not deal at all with deprecated patterns that you don't seem to be using anyway, and here I refer to the "autocommit=True" flag.  That flag is a leftover from about 12 years ago at this point, when that was the only way the Session worked (pre 0.5).   If the docs for that flag seem a little weird, you're not wrong, which is why that whole thing will be gone in 2.0.   

Next, it's important to note that what we are now called "autobegin" is for the most part not new in any way except one implementation detail, that doesn't actually impact anything:

1. the Session's notion of a "transaction" is a Python object called SessionTransaction.  When this object is created, nothing actually happens, no database connection is made.   It only makes database connections (and therefore starts real database transactions) when somethign has to actually run some SQL.  then it holds onto that connection until the SessionTransaction itself is "rolled back" or "committed"
2. in all the SQLAlchemy versions since 0.5, this SessionTransaction is made right when you construct the Session().  But as noted, no DB transaction is there.  In that sense, the Session has done "autobegin", but it's a "begin" of the transaction from the Session's point of view, not the database.
3. in 1.4, the change is, the SessionTransaction isn't made at construction time, it's made when the state of the Session itself is first altered beyond the construction.  The reason for this is nothing more than to support being able to call  "session.begin()", for frameworks and applications that prefer to work that way.
4. therefore, in all versions of SQLAlchemy since 0.5, the Session has worked from a database pov the same way; as soon as its asked to emit SQL, it starts a DB transaction, which stays until rollback or commit is called.   after that, if it's asked to emit more SQL, it starts a new transaction.  that's it

> Should I be expecting that any object graph traversals may have opened a transaction in the session to which the traversed objects belong?!

if those objects are attached to a Session, then yes.

>  In other words, is it idiomatic to do something like:

it's idiomatic to have the Session explicit throughout the whole time you are working with objects.   In the old days it looked like this:

```py
sess = Session
try:
    # < do things with sess, load objects, load attributes, etc >
   
    # < assuming data was changed, commit it >
   sess.commit()
finally:
   sess.close()
```

now we are trying to use context managers to make things more modern, this is a 1.4 only pattern:

```py
with Session(engine) as session:
    session.add(some_object)
    session.add(some_other_object)
    session.commit()

# outside the block, some_object / some_other_object are now detached
```

that example is from here: https://docs.sqlalchemy.org/en/14/orm/session_basics.html#opening-and-closing-a-session 

while some people find it challenging to make sure their application works that way, that's really all there is to it.   noting the things I mentioned about lazy loading in #7109 , you just want to make sure the state of that Session is handled.

**Comment by aathan** [comment]
I think you didn't explicitly answer whether SQLA itself ever closes transactions vs only ever opening them. I'm inferring the answer is that it does not close transactions, but instead expects that you've wrapped all operations in a session context manager or will call commit yourself at the right time. Really I'm posting this response just to make things crystal clear for readers that come to this thread later.

So, with the understanding that transactions will be left open, even when they were opened by a lazy attribute load, my next question is ... what is the emit behavior for one to many relationships -- what will mark that relationship "dirty" and in need of a re-fetch from the DB? Put another way, should I expect that every access to object.other_object will cause a query (and therefore open transaction)?

Reading the docs at https://docs.sqlalchemy.org/en/14/orm/session_state_management.html , my summary (please correct if wrong) would be that relationships are loaded when the attribute is accessed (or eagerly if configured, or if joineload etc is used in a query). Once loaded, that's it ... they are not reloaded UNTIL the session is committed, at which point SQLA assumes nothing about the state of the objects and goes back to the database for every attribute/object access.

I'm not sure I fully appreciated this fact until now: After a commit() you can expect that any access to an object will result in a refetch from the database.

right?

**Reply by CaselIT** [reply]
> I think you didn't explicitly answer whether SQLA itself ever closes transactions vs only ever opening them. I'm inferring the answer is that it does not close transactions, but instead expects that you've wrapped all operations in a session context manager or will call commit yourself at the right time. Really I'm posting this response just to make things crystal clear for readers that come to this thread later.

if the session goes out of scope the transaction will eventually get closed once the gc reclaims it. Otherwise sqlalchemy cannot know when to close a transaction.

> After a commit() you can expect that any access to an object will result in a refetch from the database.

That's the default behaviour. There is a flag on the session, that's usually suggested for asyncio, to disable this behaviour. [expire_on_commit](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=expire_on_commit#sqlalchemy.orm.Session.params.expire_on_commit)

**Reply by zzzeek** [reply]
> I think you didn't explicitly answer whether SQLA itself ever closes transactions vs only ever opening them. I'm inferring the answer is that it does not close transactions, but instead expects that you've wrapped all operations in a session context manager or will call commit yourself at the right time. Really I'm posting this response just to make things crystal clear for readers that come to this thread later.

the session closes transactions when you rollback(), commit(), close() ,  or scoped_session.remove(), or the session is garbage collected.  That last case however is not very deterministic as Python gc may or may not happen dependending on specifics, so it's best to make sure .close() or an appropriate context manager is used.

closing: https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#closing-a-session

> 
> So, with the understanding that transactions will be left open, even when they were opened by a lazy attribute load, my next question is ... what is the emit behavior for one to many relationships -- what will mark that relationship "dirty" and in need of a re-fetch from the DB? Put another way, should I expect that every access to object.other_object will cause a query (and therefore open transaction)?

no, relationships only fetch once unless you are using a dynamic relationship.   this is covered [in the tutorial here](https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#loading-relationships) as well as other places.  I will reiterate that you should make sure you spend time with the tutorial just to fill in the gaps.


> 
> Reading the docs at https://docs.sqlalchemy.org/en/14/orm/session_state_management.html , my summary (please correct if wrong) would be that relationships are loaded when the attribute is accessed (or eagerly if configured, or if joineload etc is used in a query). Once loaded, that's it ... they are not reloaded UNTIL the session is 

expired.   the .commit() operation calls expiration after the database commit, again the tutorials show why it works this way, but expiration is what makes attributes on your objects become "empty".

expiration: https://docs.sqlalchemy.org/en/14/orm/session_basics.html#expiring-refreshing


> I'm not sure I fully appreciated this fact until now: After a commit() you can expect that any access to an object will result in a refetch from the database.

expire on commit is also mentioned in the tutorial: https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#closing-a-session  and also discussed at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#committing , and probably mentioned in other places also.
